### ISSUE TITLE
Lookup and update API

### ISSUE THREAD

**Issue by rphmeier** [thrumdev/nomt#1]
There is quite a lot of logic which is orthogonal to the backend. In particular:
  * Page Layout
  * Page ID
  * Node Schema
  * Trie Traversal
  * Update logic
  * Hash Function

For these use-cases, we can decouple from the backend and work entirely in memory.

From the perspective of the traversal and update APIs, we should require that all required pages are passed into the function, i.e. they are already loaded. Determining the pages to load and loading them should be done with a backend-specific API.

Here is a very rough draft of the trie operations API.

```rust
/// Compute the needed page IDs according to the provided keys.
fn needed_page_ids(keys) -> Vec<PageId>;

/// Prove lookup for a single key
/// Fails on missing page or value mismatch.
fn prove_lookup(root, key, expected_value_hash, &PageMap) -> Result<Proof, LookupError>;

/// Prove lookup for multiple keys.
/// Fails on missing page or value mismatch.
fn prove_multi_lookup(root, [(key, expected_value_hash)], &PageMap) -> Result<Proof, LookupError>;

/// Update the trie with new key/value pairs.
/// Produces a set of pages to be altered.
/// Fails on missing page.
fn update(root, [(key, value_hash)], &PageMap) -> Result<PageMap, LookupError>;
```

One important note is that we likely need to be generic over hash function.

**Comment by gabriele-0201** [comment]
You talked about different types of backends. It's not clear to me whether those backends should be aware of the concept of pages or not. 

An important distinction we should make is: should each backend be aware of what a Page and PageId are? Should the backend only be responsible for storing and efficiently providing those pages?

If yes, then the only API needed by nomt-core is something to fetch pages on demand. They will be moved into the core, and as soon as a batch of updates is done, all modified pages will be returned to the backend, whose mandate is now to update them. In this scenario, all the logic of traversing and managing pages is handled by the core; even the translation between key-path and PageIds could be kept in the core.

The advantage of this approach is the ease of implementation, where each backend will only need to do a few tasks while most of the logic is in the core. The disadvantage is the limited types of backends that could be implemented compared to what was proposed previously.

I don't see great benefits to implement backends similar to node_hash - node just to benchmark internally. I think we should focus on paging and benchmark it against other existing databases instead

**Comment by rphmeier** [comment]
> You talked about different types of backends. It's not clear to me whether those backends should be aware of the concept of pages or not.

> An important distinction we should make is: should each backend be aware of what a Page and PageId are? Should the backend only be responsible for storing and efficiently providing those pages?

This is an important point. I believe that the generic trie operations should be completely agnostic of the storage format. This is for one crucial reason: **we should support proof verification in storage-constrained environments**. We are targeting two such environments: Polkadot PVF and Risc0 circuits.

The other reason I suggest this is that it modularizes the code and reduces a dependency. This is, of the two, the weaker reason.

There is a case to limit this just to the verification routines, so that the proving and update routines operate on pages. That is likely to be faster, but probably not substantially, as this is mostly an I/O workload.

> I don't see great benefits to implement backends similar to node_hash - node just to benchmark internally. I think we should focus on paging and benchmark it against other existing databases instead

Agreed

**Comment by pepyakin** [comment]
I don't think I have a good grasp of the amounts of logic required for every path and every use-case, but I have a gut feeling, that it may be beneficial to split up those paths. One implementation would focus on the prover case (be it zk or collator), the other implementation would concern itself only with verification. That could make both implementations more straightforward and more efficient. On top of that the proof is a stable interface that we have to upkeep anyway.

Once again, not entirely sure how much sense it makes. It all depends on the amounts of duplication it'd yield and how sizeable the benefits are, if any. Just food for thought for now.

**Comment by rphmeier** [comment]
Even within verification I could expect some benefit in having separate implementations, APIs, and assumptions based on the platform where verification is done.

The maybe counter-intuitive thing is that the verification code is what's usually run during the proving step: generating the proof is done in an arbitrary, not resource or architecture-constrained environment. But validating the proof is done inside of a ZK circuit or PVF or mobile device. That makes me think the general thrust of your point is accurate but slightly misplaced.

In any case, I don't personally have the information to decide what makes a ZK-optimized verification routine at this stage. Are we OK with the approach in #22 which provides basic implementations of both?

**Comment by rphmeier** [comment]
We had a longer conversation about the cursor API and backend independence where we discussed the expected usage of the library.

## Goals of Lookup and Update

Let's summarize the goals from that conversation:
  1. **Separate page traversal from trie traversal**. I don't mind if `Cursor` is a concrete type that uses pages under the hood, or if it's a trait that in theory allows for different types of backends. Supporting different types of backends is a non-goal, so it only really helps in tests.
  2. **Lightweight single-key traverse API**. We need to check that a set of loaded pages are sufficient to traverse and read siblings for the keys we are updating and proving. Information should be provided on where the query fails, which in turn will be used to load more pages.
  3. **Batch update**. Given a set of keys, a set of updates to those keys, and a set of pages which encompass all sibling-traversals for those keys, perform the trie update calculation and produce new pages. Leaving proving to the side for now, this may be done here as well.

## Context: Session Flow

This aligns with our intention of operating the DB as a series of sequential **sessions**, which have three phases:
  1. The Run Phase: a computation is running which performs reads from the DB. We read values synchronously from the flat store and schedule the corresponding pages to be loaded in the Read phase. The Run phase concludes when the underlying computation concludes with a set of intended changes to key-value pairs. All keys in that set which were not previously read are passed to the read phase.
  2. The Read Phase: starts at the same time as the run phase and runs in parallel with it. Continually loads required pages for keys, tests whether those pages are sufficient with the single-key traverse API, fetches key preimages, and so on. This runs until all pages and other data for all keys from the Run phase are read. The Read phase can never conclude before the Run phase.
  3. The Closing Phase: starts when the Read phase concludes. Uses the pages and preimages from the Read phase to perform a root update calculation, prepare a diff for pages, and construct merkle proofs.

Timeline of a session:
![image](https://github.com/thrumdev/nomt/assets/10121380/75a00f0e-cd90-422b-a798-863fc4a6a58b)

Run Phase / Read Phase interaction
![image](https://github.com/thrumdev/nomt/assets/10121380/7399ab92-6b12-4680-bebd-7730730a80c1)

Read Phase interacting with page DB + write phase
![image](https://github.com/thrumdev/nomt/assets/10121380/f7f5d17b-f43b-4c34-86e2-3e060aea7e55)
